### PR TITLE
Extend options module

### PR DESCRIPTION
**Pull request by Nycto**
This adds a few convenience methods to the `options` module:

* `get` with a default value
* `each`
* `map`
* `filter`
* `$`

### CODE REVIEW

**Comment by Varriount** [comment]
I'm not so sure about some of these. The `get` and `$` additions are fine, but `each`, `map`, and `filter` are a tad confusing on their behavior, since such procedures are usually expected to work on sequences of things, not single values.

**Comment by kirbyfan64** [comment]
I've never heard of `filter` on an option type, but `map` [is](http://felix-lang.org/$/usr/local/lib/felix/felix-latest/share/lib/std/datatype/option.flx#line75) [very](https://wiki.haskell.org/Monads_as_containers), [very](http://ocaml-lib.sourceforge.net/doc/Option.html) [common](https://doc.rust-lang.org/std/option/enum.Option.html#method.map).

**Comment by Nycto** [comment]
An Option is really just a sequence with a max length of 1. When you think about it like that, these functions make sense.

A few examples of `filter` on an Option:
https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#filter-java.util.function.Predicate-
http://www.scala-lang.org/api/2.11.5/index.html#scala.Option@filter(p:A=>Boolean):Option[A]

These are functions are fairly standard fair. And I've even left out a few; `flatMap`, for example.

File: lib/pure/options.nim [hunk 1]
```diff
@@ -108,6 +108,41 @@ proc get*[T](self: Option[T]): T =
...
 
+proc get*[T](self: Option[T], otherwise: T): T =
+  ## Returns the contents of this option or `otherwise` if the option is none.
+  if self.isSome:
+    self.val
+  else:
+    otherwise
+
+
+proc each*[T](self: Option[T], callback: proc (input: T): void) =
```
Comment (dom96) [inline lib/pure/options.nim]: Remove this ``void`` it's redundant.

File: lib/pure/options.nim [hunk 2]
```diff
@@ -108,6 +108,41 @@ proc get*[T](self: Option[T]): T =
...
+  ## Applies a callback to the value in this Option and returns an option
```
Comment (dom96) [inline lib/pure/options.nim]: It doesn't look like it's returning anything?

File: lib/pure/options.nim [hunk 3]
```diff
@@ -108,6 +108,41 @@ proc get*[T](self: Option[T]): T =
...
+proc filter*[T](self: Option[T], callback: proc (input: T): bool): Option[T] =
+  ## Applies a callback to the value in this Option. If the callback returns
+  ## `true`, the option is returned as a Some. If it returns false, it is
+  ## returned as a None.
+  if self.has and not callback(self.val):
+    none(T)
+  else:
+    self
+
+iterator items*[T]( self: Option[T] ): T =
```
Comment (dom96) [inline lib/pure/options.nim]: Just a little nitpick, but no spaces before/after parenthesis please.

**Comment by dom96** [comment]
As for ``filter``, @Nycto do you have a use case for it? I know it exists in Scala's library but it may be a good idea to only add the most used functionality right now.

**Comment by Nycto** [comment]
I don't have any strong use cases for `filter`. I was just defaulting to the each/map/filter combo. I'm fine with dropping it.

Another route is swapping it out with `proc flatMap`, like this:

```nimrod
proc flatMap*[T](self: Option[T], callback: proc (input: T): Option[T]): Option[T] =
  if self.has:
    callback(self.val)
  else:
    self
```

It's more versatile and can be used to filter in a pinch. It also appears to be more common than `filter` is across the other languages linked to in this thread. Thoughts?

**Comment by vegansk** [comment]
The use case for filter :-)

```nimrod
type Bar = discard
type Girl = object
  age: int
proc findNearestGirl(b: Bar): Option[Girl] = discard

var bar: Bar
bar.findNearestGirl().filter((g: Girl) => g.age >= 21)
```

File: lib/pure/options.nim (hunk 2 above)
Comment (Nycto) [inline lib/pure/options.nim]: Yup. The purpose here is to apply a callback to the contents

**Comment by Nycto** [comment]
@dom96: I pushed an updated commit with the tweaks you suggested. I left `filter` in place, per the use case @vegansk provided

File: lib/pure/options.nim [hunk 4]
```diff
@@ -108,6 +108,41 @@ proc get*[T](self: Option[T]): T =
...
 
+proc get*[T](self: Option[T], otherwise: T): T =
+  ## Returns the contents of this option or `otherwise` if the option is none.
+  if self.isSome:
+    self.val
+  else:
+    otherwise
+
+
+proc each*[T](self: Option[T], callback: proc (input: T)) =
```
Comment (dom96) [inline lib/pure/options.nim]: This still does not return an ``Option``?

Comment (dom96) [inline lib/pure/options.nim]: What is the purpose of this? It looks like ``map`` already satisfies its purpose.

File: lib/pure/options.nim [hunk 5]
```diff
@@ -108,6 +108,41 @@ proc get*[T](self: Option[T]): T =
...
+proc filter*[T](self: Option[T], callback: proc (input: T): bool): Option[T] =
+  ## Applies a callback to the value in this Option. If the callback returns
+  ## `true`, the option is returned as a Some. If it returns false, it is
+  ## returned as a None.
+  if self.has and not callback(self.val):
+    none(T)
+  else:
+    self
+
+iterator items*[T](self: Option[T]): T =
```
Comment (dom96) [inline lib/pure/options.nim]: I really don't see why this should be here. Remove it please.

Comment (vegansk) [inline lib/pure/options.nim]: @dom96, It's the standard functions for Option type infunctional programming. See filter, map here for example: http://www.scala-lang.org/api/current/index.html#scala.Option

Comment (dom96) [inline lib/pure/options.nim]: Will you really be writing:

```nim
var opt = getMaybe()
for i in opt:
  echo("OK: ", i)
```
?

Comment (vegansk) [inline lib/pure/options.nim]: Option type can be used as list with length from zero to one. So in generic programming it can be usefull.

**Comment by Araq** [comment]
The items iterator could turn out to be error-prone though:

```nim
# semantics changed silently:
for x in usedToReturnListNowReturnsOptionOfList(y):
  echo x
```

File: lib/pure/options.nim (hunk 4 above)
Comment (Nycto) [inline lib/pure/options.nim]: This is more akin to a `for` loop than a map. A `map` comes with the overhead of allocating a return value, while an `each` does not. It also means consumers don't need to discard the result.

**Comment by Nycto** [comment]
> The items iterator could turn out to be error-prone though:

This could really be said about anything that implements `items`, though. Even `seq`:

```nimrod
for x in usedToReturnListNowReturnsListOfList(y):
  echo x
```

One other thing to note about the `items` function is that it allows you to pass an Option to templates that expect a `seq`. The templates in `sequtils`, for example

File: lib/pure/options.nim (hunk 4 above)
Comment (Nycto) [inline lib/pure/options.nim]: I was originally against using `map` to perform an `each`, as I see them as separate functionalities. However, it looks like the precedent has already been set to have `map` do what an `each` does:

https://github.com/nim-lang/Nim/blob/devel/lib/system.nim#L2124-L2137

I'm not going to lie, this makes me itch. A `map` is an operation that produces a result, while an `each` operates without producing result. However, I think consistency within the core library trumps most things, so I'm willing to concede on this.

**Comment by Nycto** [comment]
New version pushed:
* Renames `each` to `map`, as that [precedent already exists in the core library](https://github.com/nim-lang/Nim/blob/371470e73f2c79f4d9babc86f59b5d580defb050/lib/system.nim#L2124-L2137).
* Removes the `items` iterator. The other functionality here is important enough (imo) to get included that I don't think it's worth debating a minor feature like that one.

File: lib/pure/options.nim [hunk 6]
```diff
@@ -108,6 +108,36 @@ proc get*[T](self: Option[T]): T =
...
 
+proc get*[T](self: Option[T], otherwise: T): T =
+  ## Returns the contents of this option or `otherwise` if the option is none.
+  if self.isSome:
+    self.val
+  else:
+    otherwise
+
+
+proc map*[T](self: Option[T], callback: proc (input: T)) =
```
Comment (dom96) [inline lib/pure/options.nim]: Would it make sense to allow in-place modification of ``self``'s value here, i.e. ``proc (input: var T)``?

**Comment by dom96** [comment]
Decided to merge as-is. But I think in-place ``map`` may have merit?

**Comment by dom96** [comment]
In any case. Thank you @Nycto and sorry about how long this took.

**Pull request merged**